### PR TITLE
Store LKI about what a permanent was attacking/blocking; fix Gomazoa

### PR DESCRIPTION
**Pull request by xenohedron**
Was investigating one of the items in #13774 and turns out [[Gomazoa]] has a unique sort of effect. I added tests for a variety of edge cases. Although I tried various approaches with new/existing watchers, no easy combination of watched events was going to work in all my test cases. My conclusion is that the info needed to be stored in the LKI of the permanent, and after implementing that functionality it works nicely.

Previously, a permanent only stored a boolean for whether it was attacking. Now, it stores a MageObjectReference referring to what it's attacking (not just UUID, because e.g. planeswalker could be blinked). Null for not attacking, so testing for null reproduces existing behavior. Added a line to update the reference in the edge case of reselecting defender. This reference isn't actually used yet by any card I know of but since I was adding the functionality for blockers I thought it was good to make it consistent.

A permanent also stores an integer for the number of creatures it's blocking, with 0 for not blocking. This is used in some weird ways to check if further blocks are possible and didn't seem to be updated on every usage I would need, and I didn't want to change existing behavior. So I added a separate set of MageObjectReference to refer to the things it's blocking. This needed a special check for when creatures are explicitly removed from combat. It still holds references to permanents that have left the battlefield, which is what we need for LKI usage. It's not for the purpose of counting, that's handled by existing watchers.

The existing combat code is pretty messy and could likely be cleaned up with some investigation, but I kept these adjustments minimal here. All tests pass.

### CODE REVIEW

File: Mage/src/main/java/mage/game/combat/Combat.java [hunk 1]
```diff
@@ -1623,8 +1623,12 @@ public boolean removeFromCombat(UUID creatureId, Game game, boolean withEvent) {
         }
         boolean result = false;
         if (withEvent) {
-            creature.setAttacking(false);
-            creature.setBlocking(0);
+            creature.setAttacking(null);
+            creature.clearBlocking();
```
Comment (JayDi85) [inline Mage/src/main/java/mage/game/combat/Combat.java:1627]: it's better to use same naming for attack-block like:
* setAttacking
* setBlocking
* clearAttacking
* clearBlocking
* and other

File: Mage/src/main/java/mage/game/combat/Combat.java [hunk 2]
```diff
@@ -1653,16 +1657,16 @@ public void endCombat(Game game) {
             for (UUID attacker : group.attackers) {
                 creature = game.getPermanent(attacker);
                 if (creature != null) {
-                    creature.setAttacking(false);
-                    creature.setBlocking(0);
+                    creature.setAttacking(null);
+                    creature.clearBlocking();
                     creature.clearBandedCards();
```
Comment (JayDi85) [inline Mage/src/main/java/mage/game/combat/Combat.java:1662]: Miss getBlockers().stream() code in other places with block clear?

Image

If miss then it must be replaced by other clear code logic due mor and mor's compare implementation (it's will keep blinked/remove permanentns forever in the list and cause error logs in some use cases, e.g. copies).

Image

File: Mage/src/main/java/mage/game/permanent/PermanentImpl.java [hunk 3]
```diff
@@ -1630,15 +1632,41 @@ public boolean canUseActivatedAbilities(Game game) {
...
     }
 
+    @Override
+    public void addBlocking(UUID attackerId, Game game) {
+        this.blockingSet.add(new MageObjectReference(attackerId, game));
+    }
+
+    @Override
+    public void removeBlocking(UUID attackerId, Game game) {
+        this.blockingSet.remove(new MageObjectReference(attackerId, game));
```
Comment (JayDi85) [inline Mage/src/main/java/mage/game/permanent/PermanentImpl.java:1656]: See above comments about clear code -- make sure it's called for existing attacker. Maybe some runtime check can help like:

``` java
if (this.blockingSet.stream().anyMatch(ref -> Objects.equals(ref.getSourceId(), attackerId))) {
  throw new IllegalArgumentException("Wrong code usage: trying to clear non existing blocking object");
}
```

**Comment by JayDi85** [comment]
[[Gomazoa]]

File: Mage/src/main/java/mage/game/combat/Combat.java (hunk 1 above)
Comment (xenohedron) [reply Mage/src/main/java/mage/game/combat/Combat.java:1627]: In general I agree about aligning method naming, but it doesn't quite make sense in this case because the underlying info stored is different. The original methods before this PR were different too. If you really want, can make `clearAttacking()` to replace `setAttacking(null)`, though I think it's adequate as is.

File: Mage/src/main/java/mage/game/permanent/PermanentImpl.java (hunk 3 above)
Comment (xenohedron) [reply Mage/src/main/java/mage/game/permanent/PermanentImpl.java:1656]: The only usage is in `removeFromCombat` with event, which as far as I can tell is only called for existing attacker. `removeFromBattlefield` calls without event.

File: Mage/src/main/java/mage/game/combat/Combat.java (hunk 2 above)
Comment (xenohedron) [reply Mage/src/main/java/mage/game/combat/Combat.java:1662]: No, getBlockers().stream() code is to correspond to removing an attacker from combat, not a blocker. It is only needed in the one place when an attacker is explicitly removed from combat. The other usages of `setAttacking(null)` are at the end of combat where everything is cleared, and during attack restriction check where nothing can be blocking yet. (Found one bad card implementation Observed Stasis which calls wrong method, will fix.) Note that because all is cleared at end of combat, will not be kept forever in any case.
